eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+    """
+    activation_layer = get_activation_layer(estimator, layer)
+    predicted = get_target_prediction(estimator, doc, targets)
+    
+    heatmap = grad_cam(estimator, doc, predicted, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations' (the output is not yet a heat map)
+
+    # need to insert a 'channel' axis to get a rank 3 image from rank 2
+    heatmap = np.expand_dims(heatmap, axis=-1)
+    heatmap = array_to_img(heatmap) # -> PIL image
+    
+    # take the single image from the input 'batch' (rank 4 to rank 3)
+    doc = doc[0]
+    image = array_to_img(doc) # -> PIL image
+
+    # TODO: return arrays, not images (arrays are more general)
+    return Explanation(
+        estimator.name, # might want to replace this with something else, eg: estimator.summary()
+        description=DESCRIPTION_KERAS,
+        error='',
+        method='Vanilla Grad-CAM',
+        is_regression=False, # TODO: classification vs regression model
+        highlight_spaces=None, # might be relevant later when explaining text models
+        image=image,
+        heatmap=heatmap,
+    )
+
+
+def get_activation_layer(estimator, layer):
+    """
+    Get an instance of the desired activation layer.
+
+    See :func:`explain_prediction_keras` for description of ``estimator`` and
+    ``layer`` parameters.
+
+    Returns
+    -------
+    A ``keras.layers.Layer`` instance.
+
+    Raises
+    ------
+    TypeError : ``layer`` is wrong type.
+    """        
+    if layer is None:
+        # Automatically get the layer if not provided
+        activation_layer = search_layer_backwards(estimator, is_suitable_activation_layer)
+    elif isinstance(layer, Layer):
+        activation_layer = layer
+    elif isinstance(layer, int):
+        # bottom-up horizontal graph traversal
+        activation_layer = estimator.get_layer(index=layer)
+        # These can raise ValueError if the layer index / name specified is not found
+    elif isinstance(layer, str):
+        activation_layer = estimator.get_layer(name=layer)
+    else:
+        raise TypeError('Invalid layer (must be str, int, keras.layers.Layer, or None): %s' % layer)
+
+    # TODO: validate activation_layer dimensions (is it possible to perform Grad-CAM on it?)
+    return activation_layer
+
+
+def search_layer_backwards(estimator, condition):
+    """
+    Search for a layer in ``estimator`` backwards (starting from output layer),
+    checking if the layer is suitable with the callable ``condition``,
+    where condition takes ``estimator`` and ``index`` arguments.
+    
+    Returns
+    -------
+    layer : object
+        A suitable ``keras.layers.Layer`` instance.
+
+    Raises
+    ------
+    ValueError :
+        If suitable layer can not be found.
+    """
+    # we assume that this is a simple feedforward network
+    # linear search in reverse
+    i = len(estimator.layers)-1
+    while -1 < i and not condition(estimator, i):
```

Comment:
I think ``i >= 0`` would be a more common way to write this condition

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/2821b25830a7dbca1dccbfa64f8e92a166908b6a